crypto/evp/keymgmt_lib.c

```diff
@@ -0,0 +1,307 @@
+            return pk->pkeys[i].provkey;
+    }
+
+    if (pk->pkey.ptr != NULL) {
+        /* There is a legacy key, try to export that one to the provider */
+
+        /* If the legacy key doesn't have an export function, give up */
+        if (pk->ameth->export_to == NULL)
+            return NULL;
+
+        /* Otherwise, simply use it */
+        provkey = pk->ameth->export_to(pk, keymgmt);
+    } else {
+        /*
+         * Here, there is no legacy key, so we look at the already
+         * cached but there are cached provider keys, and import from
+         * the first that supports it (i.e. use its export function),
+         * and export the imported data to the new provider.
+         */
+
+        /*
+         * If the given keymgmt doesn't have an import function, give up
+         */
+        if (keymgmt->importkey_priv == NULL)
+            return NULL;
+
+        for (j = 0; j < i && pk->pkeys[j].keymgmt != NULL; j++) {
+            if (pk->pkeys[j].keymgmt->exportkey_priv != NULL) {
+                const OSSL_PARAM *paramdefs = NULL;
+                OSSL_PARAM *params = NULL;
+                void *data = NULL;
+                void *provctx =
+                    ossl_provider_ctx(EVP_KEYMGMT_provider(keymgmt));
+
+                paramdefs = pk->pkeys[j].keymgmt->exportkey_priv_types();
+                /*
+                 * All params have 'data' set to NULL.  In that case,
+                 * the exportkey call should just fill in 'return_size'
+                 * in all applicable params.
+                 */
+                params = paramdefs_to_params(paramdefs);
+                /* Get 'return_size' filled */
+                pk->pkeys[j].keymgmt->exportkey_priv(pk->pkeys[j].provkey,
+                                                     params);
+
+                /*
+                 * Allocate space and assign 'data' to point into the
+                 * data block
+                 */
+                data = allocate_params_space(params);
+
+                /*
+                 * Call the exportkey function a second time, to get
+                 * the data filled
+                 */
+                pk->pkeys[j].keymgmt->exportkey_priv(pk->pkeys[j].provkey,
+                                                     params);
+
+                /*
+                 * We should have all the data at this point, so import
+                 * into the new provider and hope to get a key back.
+                 */
+                provkey = keymgmt->importkey_priv(provctx, params);
+                OPENSSL_free(params);
+                OPENSSL_free(data);
+
+                if (provkey != NULL) {
+                    break;
+                }
+            }
+        }
+    }
+
+    /*
+     * TODO(3.0) Right now, we assume we have ample space.  We will
+     * have to think about a cache aging scheme, though, if |i| indexes
+     * outside the array.
+     */
+    j = ossl_assert(i < OSSL_NELEM(pk->pkeys));
```

Comment:
I think this should just return an error rather than asserting this. It *could* happen through normal use (although unlikely) - whereas an assert should only be for "this should never happen".

Reply:
I'm really having that in place of a proper cache ageing mechanism, so *for now*, I've left that unimplemented with the belief that we won't hit that many exports simply because there are only so many providers available, but we *should* implement a proper ageing mechanism at some point before 3.0 is released.  I see no reason to play nice in the mean time, and should this assertion be triggered before then, we'd better hustle our buns to get that ageing mechanism in place.

Reply:
ok